File: gto-support-db/src/main/java/org/ccci/gto/android/common/db/BaseContract.java
package org.ccci.gto.android.common.db;

import android.provider.BaseColumns;
import android.support.annotation.NonNull;
import android.text.TextUtils;

public abstract class BaseContract {
    public interface Base extends BaseColumns {
        String COLUMN_ROWID = _ID;
        String SQL_COLUMN_ROWID = COLUMN_ROWID + " INTEGER PRIMARY KEY";
    }

    @NonNull
    public static String uniqueIndex(@NonNull final String... columns) {
        if (columns.length < 1) {
            throw new IllegalArgumentException("There needs to be at least 1 column specified for a Unique index");
        }
        return "UNIQUE(" + TextUtils.join(",", columns) + ")";
    }

    @NonNull
    public static String create(@NonNull final String table, @NonNull final String... sqlColumns) {
        return "CREATE TABLE " + table + " (" + TextUtils.join(",", sqlColumns) + ")";
    }

    @NonNull
    public static String drop(@NonNull final String table) {
        return "DROP TABLE IF EXISTS " + table;
    }
}

File: gto-support-db/src/main/java/org/ccci/gto/android/common/db/BaseContract.kt
package org.ccci.gto.android.common.db

import android.provider.BaseColumns

abstract class BaseContract {
    interface Base : BaseColumns {
        companion object {
            const val COLUMN_ROWID = BaseColumns._ID
            const val SQL_COLUMN_ROWID = "$COLUMN_ROWID INTEGER PRIMARY KEY"
        }
    }

    companion object {
        @JvmStatic
        fun uniqueIndex(vararg columns: String): String {
            require(columns.isNotEmpty()) { "There needs to be at least 1 column specified for a Unique index" }
            return "UNIQUE(${columns.joinToString(",")})"
        }

        @JvmStatic
        fun create(table: String, vararg sqlColumns: String) = "CREATE TABLE $table (${sqlColumns.joinToString(",")})"

        @JvmStatic
        fun drop(table: String) = "DROP TABLE IF EXISTS $table"
    }
}
